File: General-DSA-Questions/Heap/7_minimumCostToConnect-N-Ropes.cpp
// Write a function to find the minimum cost to connect N ropes.

// You're given n ropes with different lengths.
// To connect two ropes, it costs sum of their lengths.
// Your task is to connect all ropes into one with minimum total cost.

// Intitution:
// Always combine the two smallest ropes first.
// This is because combining the two smallest ropes will minimize the cost of connecting them.
// We can use a min-heap to efficiently find the two smallest ropes.

// Approach:
// 1. Create a min-heap to store the lengths of the ropes.
// 2. While there is more than one rope in the heap:
//    a. Extract the two smallest ropes from the heap.
//    b. Calculate the cost of connecting them (sum of their lengths).
//    c. Add the cost to the total cost.
//    d. Insert the combined rope back into the heap.
// 3. Return the total cost.
// 4. Time Complexity: O(N log N) for inserting and extracting elements from the heap.
// 5. Space Complexity: O(N) for storing the ropes in the heap.

#include <iostream>
#include <queue>
using namespace std;

// Function to find the minimum cost to connect N ropes
int minCostToConnectRopes(int arr[], int n) {
    // Create a min-heap using priority queue
    priority_queue<int, vector<int>, greater<int>> minHeap;

    // Insert all rope lengths into the min-heap
    for (int i = 0; i < n; i++) {
        minHeap.push(arr[i]);
    }

    int totalCost = 0;

    // While there is more than one rope in the heap
    while (minHeap.size() > 1) {
        // Extract the two smallest ropes from the heap
        int first = minHeap.top();
        minHeap.pop();
        int second = minHeap.top();
        minHeap.pop();

        // Calculate the cost of connecting them
        int cost = first + second;
        totalCost += cost;

        // Insert the combined rope back into the heap
        minHeap.push(cost);
    }

    return totalCost; // Return the total cost
}

int main() {
    int arr[] = {4, 3, 2, 6}; // Example rope lengths
    int n = sizeof(arr) / sizeof(arr[0]);

    int minCost = minCostToConnectRopes(arr, n);
    cout << "Minimum cost to connect ropes: " << minCost << endl;

    return 0;
}
// Output: Minimum cost to connect ropes: 29

File: JS-DSA-Questions/Heap/7_minimumCostToConnect-N-Ropes.js
// Minimum cost to connect N ropes.
// Connecting two ropes costs the sum of their lengths. Connect all ropes into one with minimum
// total cost.

// Intuition (greedy): always combine the two SMALLEST ropes first. A min-heap gives the two
// smallest efficiently.

// Approach:
// 1. Push all lengths into a min-heap.
// 2. While more than one rope remains: pop the two smallest, add their sum to the total, push the
//    combined rope back.
// Time: O(N log N). Space: O(N).

// JavaScript note: JS has no built-in priority queue, so we include a small MinHeap class.
class MinHeap {
  constructor() {
    this.a = [];
  }
  size() {
    return this.a.length;
  }
  push(v) {
    this.a.push(v);
    let i = this.a.length - 1;
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (this.a[parent] > this.a[i]) {
        [this.a[parent], this.a[i]] = [this.a[i], this.a[parent]];
        i = parent;
      } else break;
    }
  }
  pop() {
    const top = this.a[0];
    const last = this.a.pop();
    if (this.a.length > 0) {
      this.a[0] = last;
      let i = 0;
      const n = this.a.length;
      while (true) {
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        let smallest = i;
        if (l < n && this.a[l] < this.a[smallest]) smallest = l;
        if (r < n && this.a[r] < this.a[smallest]) smallest = r;
        if (smallest !== i) {
          [this.a[i], this.a[smallest]] = [this.a[smallest], this.a[i]];
          i = smallest;
        } else break;
      }
    }
    return top;
  }
}

function minCostToConnectRopes(arr) {
  const minHeap = new MinHeap();
  for (const len of arr) minHeap.push(len);

  let totalCost = 0;
  while (minHeap.size() > 1) {
    const first = minHeap.pop();
    const second = minHeap.pop();
    const cost = first + second;
    totalCost += cost;
    minHeap.push(cost);
  }

  return totalCost;
}

console.log("Minimum cost to connect ropes:", minCostToConnectRopes([4, 3, 2, 6])); // 29
